llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -0,0 +1,150 @@
+  auto *Ty = Type::getInt32Ty(Ctx);
+  SmallVector<Metadata *, 3> MD;
+  MD.push_back(ConstantAsMetadata::get(
+      Constant::getIntegerValue(Ty, APInt(32, OpCode))));
+  MD.push_back(
+      ConstantAsMetadata::get(Constant::getIntegerValue(Ty, APInt(32, Value))));
+  MD.push_back(MDString::get(Ctx, Secondary));
+  return MDNode::get(Ctx, MD);
+}
+
+} // anonymous namespace
+
+PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
+                                                 ModuleAnalysisManager &MAM) {
+  LLVMContext &Ctx = M.getContext();
+  unsigned MDKindID = Ctx.getMDKindID(SPIRV_DECOR_MD_KIND);
+  bool AnyCompileTimePropertiesMet = false;
+
+  // Let's process all the globals
+  for (auto &GV : M.globals()) {
+    // we suppose the enumeration orders in every enumeration in the SYCL
+    // headers are the same as in the descriptions of the corresponding
+    // decorations in the SPV_INTEL_* extensions.
+    SmallVector<Metadata *, 8> MDOps;
+    for (auto &Attribute : GV.getAttributes()) {
+      // Currently, only string attributes are supported
+      if (!Attribute.isStringAttribute())
+        continue;
+      auto DecorIt = SpirvDecorMap.find(Attribute.getKindAsString());
+      if (DecorIt == SpirvDecorMap.end())
+        continue;
+      auto Decor = DecorIt->second;
+      auto DecorCode = Decor.Code;
+      auto DecorValue = Decor.Type == DecorValueTy::uint32
+                            ? getAttributeAsInteger<uint32_t>(Attribute)
+                            : hasProperty(Attribute);
+      MDOps.push_back(buildSpirvDecorMetadata(Ctx, DecorCode, DecorValue));
+    }
+
+    // Some properties should be handled specially.
+
+    // The host_access property is handled specially for device global variables
+    // because the SPIR-V decoration requires two "extra operands". The second
+    // SPIR-V operand is the "name" (the value of the "sycl-unique-id" property)
+    // of the variable.
+    if (isDeviceGlobalVariable(GV)) {
+      auto HostAccessDecorValue =
+          GV.hasAttribute(SYCL_HOST_ACCESS_ATTR)
+              ? getAttributeAsInteger<uint32_t>(GV, SYCL_HOST_ACCESS_ATTR)
+              : SPIRV_HOST_ACCESS_DEFAULT_VALUE;
+      auto VarName = getGlobalVariableUniqueId(GV);
+      MDOps.push_back(buildSpirvDecorMetadata(Ctx, SPIRV_HOST_ACCESS_DECOR,
+                                              HostAccessDecorValue, VarName));
+    }
+
+    // Add the generated metadata to the variable
+    if (!MDOps.empty()) {
+      GV.addMetadata(MDKindID, *MDNode::get(Ctx, MDOps));
+      AnyCompileTimePropertiesMet = true;
+    }
+  }
+
+  // The pass just adds some metadata to the module, it should not ruin
+  // any analysis, but we need return PreservedAnalyses::none() to inform
+  // the caller that at least one compile-time property was met.
+  return AnyCompileTimePropertiesMet ? PreservedAnalyses::none()
+                                     : PreservedAnalyses::all();
```

Comment:
Then I'd add a TODO at least to find a way to return a result without resetting analysis state.

Reply:
I found one way: just to repeat the body of the method but instead of generate any metadata just do 'or' for a boolean variable, so to check if there any globals, if a global has registered properties and so on but it requires code duplication and looks ugly. From the other point of view, currently we run the pass in isolation and it doesn't matter whether cleanup the pass any analysis. Therefore, I'm not sure if the TODO is actually required.

Reply:
Using PreservedAnalyses as a way to forward a result of the pass execution looks hacky.
But ok, let's overlook it now since current pass is run in isolation.